File: go/v1/client/signature.go
package client

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
)

const signaturePrefix = "sha256="

// Sign computes an HMAC-SHA256 signature for a PayGate API request.
// Sign string: METHOD\nPATH\nTIMESTAMP\nBODY
func Sign(apiSecret, method, path, timestamp, body string) string {
	signStr := method + "\n" + path + "\n" + timestamp + "\n" + body
	mac := hmac.New(sha256.New, []byte(apiSecret))
	mac.Write([]byte(signStr))
	return signaturePrefix + base64.StdEncoding.EncodeToString(mac.Sum(nil))
}

// Verify checks a signature against the expected value.
func Verify(secret, expectedSig, method, path, timestamp, body string) bool {
	actual := Sign(secret, method, path, timestamp, body)
	return subtle.ConstantTimeCompare([]byte(actual), []byte(expectedSig)) == 1
}

File: php/v1/src/Signature.php
<?php

namespace Paygate\Sdk;

/**
 * HMAC-SHA256 signing and verification for PayGate API requests.
 *
 * Sign string: METHOD\nPATH\nTIMESTAMP\nBODY
 * Output: sha256=<base64-hmac>
 *
 * @author alphah
 * @since 1.0.0
 */
class Signature
{
    /**
     * Compute HMAC-SHA256 signature.
     */
    public static function sign(string $apiSecret, string $method, string $path, string $timestamp, string $body): string
    {
        $signStr = "{$method}\n{$path}\n{$timestamp}\n{$body}";
        $hash = hash_hmac('sha256', $signStr, $apiSecret, true);
        return 'sha256=' . base64_encode($hash);
    }

    /**
     * Verify a signature against the expected value.
     */
    public static function verify(string $secret, string $expectedSig, string $method, string $path, string $timestamp, string $body): bool
    {
        $actual = self::sign($secret, $method, $path, $timestamp, $body);
        return hash_equals($actual, $expectedSig);
    }
}
